KG builder components user guide (#102)
neo4j/neo4j-genai-python: docs/source/user_guide_kg_builder.rst

.. _user-guide-kg-builder:

User Guide: Knowledge Graph Builder
###################################

This page provides information about how to create a Knowledge Graph from
unstructured data.

.. warning::

    This feature is still experimental. API changes and bug fixes are expected.

    It is not recommended to use it in production yet.

******************

> Pipeline structure

Reviewer: ```suggestion
Pipeline Structure
```

******************

A Knowledge Graph (KG) construction pipeline requires a few components:

> - Document **parser**: extract text from files (PDFs, ...)

Author: Maybe we should just bold the full expression: **Document parser**, **Chunk embedder**, etc?

                properties=[
                    SchemaProperty(name="name", type="STRING"),
                    SchemaProperty(name="country", type="STRING"),
                ],
            ),
        ],
        relations=[
            SchemaRelation(
                label="WORKED_ON",
            ),
            SchemaRelation(
                label="WORKED_FOR",
            ),
        ],
        possible_schema=[
            ("Person", "WORKED_ON", "Field"),
            ("Person", "WORKED_FOR", "Organization"),
        ],
    )

After validation, this schema is saved in a `SchemaConfig` object, whose dict representation is passed
to the LLM.

Entity and Relation Extractor
=============================

This component is responsible for extracting the relevant entities and relationships from each text chunk,
using the schema as guideline.

This package contains an LLM-based entity and relationships extractor: `LLMEntityRelationExtractor`.
It can be used in this way:

.. code:: python

    from neo4j_genai.experimental.components.entity_relation_extractor import (
        LLMEntityRelationExtractor,
    )
    from neo4j_genai.llm import OpenAILLM

>     extractor = LLMEntityRelationExtractor(

Reviewer: I would add a section on inserting a custom prompt / prompt template. I think most users will be very interested in that

Author: Isn't this covered here: https://github.com/neo4j/neo4j-genai-python/pull/102/files#diff-c37850688512e77f403fe5c0f70c11f569a5ae233777f9859877f2f49dfb4619R306 ?

Reviewer: Sorry I missed that!

Author: No pb, thanks for the review!

.. _user-guide-kg-builder:

User Guide: Knowledge Graph Builder
########################################

This page provides information about how to create a Knowledge Graph from
unstructured data.

.. warning::

    This feature is still experimental. API changes and bug fixes are expected.

    It is not recommended to use it in production yet.

******************************
Pipeline structure
******************************

A Knowledge Graph (KG) construction pipeline requires a few components:

- A document **parser**: extract text from files (PDFs, ...)
- A document **chunker**: split the text into smaller pieces of text, manageable by the LLM context window (token limit).

> - A chunk **embeder** (optional): compute and store the chunk embeddings

Reviewer: ```suggestion
- A chunk **embedder** (optional): compute and store the chunk embeddings
```

- A **schema builder**: provide a schema to ground the LLM extracted entities and relations and obtain an easily navigable KG.
- An **entity and relation extractor**: extract relevant entities and relations from the text.

> - A **Knowledge Graph writer**: write the identified entities and relations to a Neo4j database.

Reviewer: Maybe it's just me being very nitpicky, but I think this section is more readable if each sentence doesn't start with "A..." or "An...". I felt the same way about the other user guides

.. image:: images/kg_builder_pipeline.png
  :alt: KG Builder pipeline

This package contains the interface and implementations for each of these components, which are detailed in the following sections.

To see an end-to-end example of a Knowledge Graph construction pipeline,
refer to `this example <https://github.com/neo4j/neo4j-genai-python/blob/main/examples/pipeline/kg_builder.py>`_.

**********************************
Knowledge Graph Builder Components
**********************************

> Below is a list of the different components available in this package, and how to use them.

Reviewer: ```suggestion
Below is a list of the different components available in this package and how to use them.
```

I don't think the comma is required

Each of these components can be run individually:

.. code:: python

    import asyncio

>     # replace `PdfLoader` by a real component name

Reviewer: I'm not sure what is meant by "real component name" here

Author: Will remove it, I used a non existing component at the beginning but found it more useful to choose one real component instead.

    asyncio.run(my_component.run("my_file.pdf"))

They can also be used within a pipeline:

.. code:: python

    from neo4j_genai.experimental.pipeline import Pipeline
    # replace `PdfLoader` by a real component name
    from neo4j_genai.experimental.components.pdf_loader import PdfLoader
    pipeline = Pipeline()
    my_component = PdfLoader()
    pipeline.add("component_name", my_component)

Document Parser
===============

Document parsers start from a file path and return the text extracted from this file.

This package currently supports text extraction from PDFs:

.. code:: python

    from pathlib import Path
    from neo4j_genai.experimental.components.pdf_loader import PdfLoader

    loader = PdfLoader()
    loader.run(path=Path("my_file.pdf"))

To implement your own loader, use the `DataLoader` interface:

.. code:: python

    from pathlib import Path
    from neo4j_genai.experimental.components.pdf_loader import DataLoader, PdfDocument

    class MyDataLoader(DataLoader):
        async def run(self, path: Path) -> PdfDocument:
            # process file in `path`
            return PdfDocument(text="text")

Document Splitter
=================

> Document splitters, as the name indicate, are splitting documents into smaller chunks

Reviewer: ```suggestion
Document splitters, as the name indicates, split documents into smaller chunks
```

    class MyTextSplitter(TextSplitter):

        def __init__(self, separator: str = ".") -> None:
            self.separator = separator

        async def run(self, text: str) -> TextChunks:
             return TextChunks(
                 chunks=[
                     TextChunk(text=text_chunk)
                     for text_chunk in text.split(self.separator)
                 ]
             )

Chunk Embedder
==============

In order to embed the chunks' texts (to be used in vector search RAG), one can use the
`TextChunkEmbedder` component, which rely on the :ref:`Embedder` interface.

Example usage:

.. code:: python

    from neo4j_genai.experimental.components.embedder import TextChunkEmbedder
    from neo4j_genai.embeddings.openai import OpenAIEmbeddings
    text_chunk_embedder = TextChunkEmbedder(embedder=OpenAIEmbeddings())
    text_chunk_embedder.run(text_chunks=TextChunks(chunks=[TextChunk(text="my_text")]))

.. note::

    To use OpenAI (embedding or LLM), the `OPENAI_API_KEY` must be in the env vars, for instance using:

    .. code:: python

        import os
        os.environ["OPENAI_API_KEY"] = "sk-..."

> If OpenAI is not an option, see :ref:`embedders` to learn how to use sentence-transformer or create your own embedder.

Reviewer: ```suggestion
If OpenAI is not an option, see :ref:`embedders` to learn how to use sentence-transformers or create your own embedder.
```
